File: test-suite-kotlin-ksp/src/test/kotlin/io/micronaut/docs/http/server/bind/type/ShoppingCartRequestArgumentBinder.kt
package io.micronaut.docs.http.server.bind.type

// tag::class[]
import io.micronaut.core.bind.ArgumentBinder
import io.micronaut.core.bind.ArgumentBinder.BindingResult
import io.micronaut.core.convert.ArgumentConversionContext
import io.micronaut.core.type.Argument
import io.micronaut.http.HttpRequest
import io.micronaut.http.bind.binders.TypedRequestArgumentBinder
import io.micronaut.jackson.serialize.JacksonObjectSerializer
import java.util.Optional
import jakarta.inject.Singleton

@Singleton
class ShoppingCartRequestArgumentBinder(private val objectSerializer: JacksonObjectSerializer) :
    TypedRequestArgumentBinder<ShoppingCart> {

    override fun bind(
        context: ArgumentConversionContext<ShoppingCart>,
        source: HttpRequest<*>
    ): BindingResult<ShoppingCart> { //<1>

        val cookie = source.cookies["shoppingCart"]

        return if (cookie == null)
            BindingResult {
                Optional.empty()
            }
        else {
            BindingResult {
                objectSerializer.deserialize( // <2>
                    cookie.value.toByteArray(),
                    ShoppingCart::class.java
                )
            }
        }
    }

    override fun argumentType(): Argument<ShoppingCart> {
        return Argument.of(ShoppingCart::class.java) //<3>
    }
}
// end::class[]

File: test-suite-python/src/test/python/micronaut/docs/http/server/bind/type/ShoppingCartRequestArgumentBinder.py
# tag::class[]
import java
from jakarta.inject import Singleton
from micronaut.core.bind import ArgumentBinder
from micronaut.core.convert import ArgumentConversionContext
from micronaut.core.type import Argument
from micronaut.http import HttpRequest
from micronaut.http.bind.binders import TypedRequestArgumentBinder
from micronaut.jackson.serialize import JacksonObjectSerializer

ShoppingCartClass = java.type("micronaut.docs.http.server.bind.type.ShoppingCart")


@Singleton
class ShoppingCartRequestArgumentBinder(TypedRequestArgumentBinder):
    def __init__(self, objectSerializer: JacksonObjectSerializer):
        self.objectSerializer = objectSerializer

    def bind(self, context: ArgumentConversionContext, source: HttpRequest):  # <1>
        cookie = source.getCookies().get("shoppingCart")
        if cookie is None:
            return ArgumentBinder.BindingResult.empty()

        return lambda: self.objectSerializer.deserialize(  # <2>
            bytearray(cookie.getValue(), "utf-8"),
            ShoppingCartClass,
        )

    def argumentType(self):
        return Argument.of(ShoppingCartClass)  # <3>
# end::class[]
